File: AndroidStudioProjects/MethodsAndClasses/app/src/main/java/com/ecrinaybay/methodsandclasses/MainActivity.java
package com.ecrinaybay.methodsandclasses;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;

public class MainActivity extends AppCompatActivity {

    String username; //Global değişken olduğu için farklı metotların içinden ulaşabildik.

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        System.out.println("on create called");
        testMethod();

        System.out.println(math(3,5));

        System.out.println(newMethod("Ecrin"));

        username = "";

        makeMusicians();
        makeSimpsons();

    }

    public void makeSimpsons(){
        Simpsons homer = new Simpsons("Homer",50,"Nuclear");
        /*homer.age = 51; //get ve set yoksa private yaptığımızda erişemeyiz.
        System.out.println(homer.age);*/

        System.out.println(homer.getName());
        homer.setName("Homer Simpsons"); //set ile alarak değişiklik yaptık ve yazdırdık
        System.out.println(homer.getName()); //get ile alarak yazdırdık

    }

    public void makeMusicians(){
        //instance

        Musicians james = new Musicians("James","Guitar",50);
        System.out.println(james.instrument);

    }


    public void testMethod() {
        //String username = "Lark"; //Buradaki username globalde tanımladığımız usernameden farklıdır.
        username = "Lark";
        int x = 4 + 4;
        x = 9;
        System.out.println("value of x" + x);
    }

    public int math(int a, int b){
        username = "Kirk"; //Globaldeki username
        int x = 11;
        System.out.println("value of x in math: " + x);
        return a + b;
    }

    public String newMethod(String string){
        username = "Rob";
        return string + " new Method";
    }


    @Override
    protected void onResume() {
        super.onResume();
        System.out.println("on resume called");
    }

    @Override
    protected void onStop() {
        super.onStop();
        System.out.println("on stop called");
    }

    @Override
    protected void onPause() {
        super.onPause();
        System.out.println("on pause called");
    }
}

File: AndroidStudioProjects/KotlinDelegates/app/src/main/java/com/ecrinaybay/kotlindelegates/MainActivity.kt
package com.ecrinaybay.kotlindelegates

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import androidx.lifecycle.LifecycleOwner

class MainActivity : AppCompatActivity(), LifecycleLogger by LifecycleLoggerImplementation() {

    //property delegates
    //lazy ile buna ulaşılana kadar hafızada yer kaplamıyor
    private val myVariable by lazy {
        println("hello this is a lazy implementation")
        10
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        registerLifecycleOwner(this)
        println(myVariable)
    }

}

interface LifecycleLogger{
    fun registerLifecycleOwner(owner: LifecycleOwner)
}

class LifecycleLoggerImplementation : LifecycleLogger, LifecycleEventObserver{
    override fun registerLifecycleOwner(owner: LifecycleOwner) {
        owner.lifecycle.addObserver(this)
    }

    override fun onStateChanged(source: LifecycleOwner, event: Lifecycle.Event) {
        when(event){
            Lifecycle.Event.ON_RESUME -> println("on resume executed")
            Lifecycle.Event.ON_PAUSE -> println("on pause executed")
            else -> Unit
        }
    }
}























































